### PR TITLE
[balsa] Call ParserCallback::onMessageBegin() early.

### PR DESCRIPTION
**Pull request by bencebeky**
When the first byte of the message arrives, check if it is the first character of a valid method name for a request, or if it is 'H' for a response (responses must start with "HTTP").  If so, call ParserCallback::onMessageBegin() immeditately.  If not, signal an error immediately.

This is already the behavior of http-parser.

Add three tests.  Unfortunately the response starting with 'H' case cannot be easily tested, as ParserCallbacks are not directly accessible, and ClientConnectionImpl::onMessageBegin() does not have any relevant side effects that can be easily tested.

Also add comment referencing #18819.  A way to disable method name validation will be needed in the future.

Tracking issue: https://github.com/envoyproxy/envoy/issues/21245

Signed-off-by: Bence Béky <bnc@google.com>

Commit Message: [balsa] Call ParserCallback::onMessageBegin() early.
Additional Description:
Risk Level: low, changed code protected by existing default-false runtime flag.
Testing: //test/common/http/http1:codec_impl_test, //test/integration:idle_timeout_integration_test
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
Runtime guard: envoy.reloadable_features.http1_use_balsa_parser

### CODE REVIEW

File: source/common/http/http1/balsa_parser.cc [hunk 1]
```diff
@@ -268,6 +268,7 @@ void BalsaParser::MessageDone() {
   }
   status_ = convertResult(connection_->onMessageComplete());
   framer_.Reset();
+  on_message_begin_called_ = false;
```
Comment (diannahu) [inline source/common/http/http1/balsa_parser.cc:271]: completely out of scope for this change, okay to consider later: Would it be right to reset other members also, e.g., headers_done_ and/or others? IIUC, resetting on_message_begin_called_ sort of suggests that we can reuse BalsaParser for a subsequent message (hm, or maybe just subsequent calls to execute()?).

**Review by diannahu — APPROVED** [review]
A nice way to begin execution!

File: source/common/http/http1/balsa_parser.cc (hunk 1 above)
Comment (bencebeky) [reply source/common/http/http1/balsa_parser.cc:271]: Yes, I think that makes perfect sense, I'll look into it.

**Comment by birenroy** [comment]
LGTM. Let me know when you'd like us to take another look.

**Review by KBaichoo — APPROVED** [review]
Looks good, thank you!

**Comment by bencebeky** [comment]
Test failure still looks unrelated.

/retest

File: source/common/http/http1/balsa_parser.cc (hunk 1 above)
Comment (bencebeky) [reply source/common/http/http1/balsa_parser.cc]: Thanks for catching this!  I filed #24557 for resetting `headers_done_`.  I couldn't identify other members that would need to be reset.  I'm fixing this, among other things, with test coverage for plenty of weird cases, at #24594.

**Pull request merged**